### DISCUSSION TITLE
another circadian rithm in pyscript

### DISCUSSION THREAD

**Discussion by phsdv** [custom-components/pyscript#499] [Show and tell]
There are already many circadian scripts, custom components and blueprints etc, but I like to have my lamps bright and cold white in the morning and warm in the evening and dimmed down at night. The lights should follow my schedule, not the sun schedule, as the sun changes too much from winter to summer. It is based on something similar in yaml (see here: [https://community.home-assistant.io/t/diy-circadian-lighting-no-flux-and-no-custom-components/179047/14](community.home-assistant.io))
You have to define your schedule in a list called BrightColorTemp. In my example, if the time is between 9:00 and 16:45 the brightness would be set at the maximum value (255) and the colortemperature at 4500 Kelvin. All lamps that should follow the rithm should be listed in CircadianLights.

Improvements to be made: 

1. the CircadianLights list should be a group light from HA. so that we have a more central place to define which lamps are having a circadian rithm.
2. When a light is switched on, it can take max 5 minutes (or whatever time you specify in the cron statement) before it gets updates. So a new function needs to be made that triggers when any of the lights in the group is switched on and then sets the values for brightness and color temperature at that moment.

I hope it is useful to someone else.
[Edit1] replaced sensor.time with template.now(hass). Thanks @ALERTua 

```python
from homeassistant.helpers import template

TransistionTime = 60
# group of ligths that should follow the circadian rithm
CircadianLights =["light.bulb_1", "light.bulb_2", "light.another_light", "light.light_5",  "light.kitchen", "light.bedroom"]
  
# [[time1, brightnes1, colortemp1], [time2, brightnes2, colortemp2], ....]
# time as int in hhmm format. no leading zeros, thus 0 means hour=0,minutes=0
# 0 means 0:00, 630 means 6:30, 2359 means 23:59
# first element should always start with time=0
BrightColorTemp = [
    [0, 1, 1800],
    [615, 10, 2000], [630, 100, 3750], [645, 200, 3750], [655, 255, 4500],
    [700, 255, 5000],
    [800, 255, 5000], [815,255,4833], [830,255,4666], [845,255,4500],
    [900, 255, 4500],
    [1645, 255, 4000],
    [1700, 255, 3500], [1715, 255, 3300], [1730, 255, 3000], [1745, 255, 2900], [1750, 255, 2800], [1755, 255, 2700],
    [1800, 255, 2600], [1815, 255, 2600], [1830, 255, 2600], [1845,254, 2500],
    [1900, 250, 2250], [1930, 245, 2000],
    [2000, 240, 1900], [2030, 230, 1850],
    [2100, 100, 1800], 
    [2200, 50, 1800], 
    [2300, 1, 1800]
    ]

# trigger every 5 minutes is circadian_on is true
@time_trigger("cron(*/5 * * * *)")
@state_active("input_boolean.circadian_on == 'on'")
def circadian():
    """ run every 5 minutes 
        and adjust brightness and color temperature for lamps that are switched on
    """

    now = template.now(hass).hour *100 + template.now(hass).minute
    x = [i for i, y in enumerate(BrightColorTemp) if now >= y[0]]
    t, br, T = BrightColorTemp[x[-1]]

    for id in CircadianLights:
        if state.get(id) == 'on':
            light.turn_on(entity_id=id, brightness=br, color_temp_kelvin=T, transition=TransistionTime)
            log.info(f"{now}: Circadian: {id}, {br}, {T}K")
```

**Comment by ALERTua** [comment]
fyi you can get `datetime.now` object without sensors and string manipulation using
```
from homeassistant.helpers import template
now = template.now(hass)
```
via https://github.com/ALERTua/ha_pyscript_modules/blob/main/entities/ha.py
